tools/codegen/api/python.py

```diff
@@ -0,0 +1,927 @@
+#
+#                     Python / C++ Args Binding
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+UNPACK_METHODS = {
+    'const Tensor &': 'tensor',
+    'Tensor &': 'tensor',
+    'Stream': 'stream',
+    'c10::optional<Tensor>': 'optionalTensor',
+    'const c10::optional<Tensor>&': 'optionalTensor',
+    'c10::optional<Generator>': 'generator',
+    'Storage': 'storage',
+    'Storage &': 'storage',
+    'const ScalarType &': 'scalartype',
+    'const Device &': 'device',
+    'c10::optional<DimnameList>': 'toDimnameListOptional',
+    'c10::optional<ScalarType>': 'scalartypeOptional',
+    'c10::optional<Layout>': 'layoutOptional',
+    'c10::optional<MemoryFormat>': 'memoryformatOptional',
+    'c10::optional<Scalar>': 'scalarOptional',
+    'c10::optional<IntArrayRef>': 'intlistOptional',
+    'c10::optional<int64_t>': 'toInt64Optional',
+    'c10::optional<bool>': 'toBoolOptional',
+    'c10::optional<double>': 'toDoubleOptional',
+    'c10::optional<ArrayRef<double>>': 'doublelistOptional',
+    'ArrayRef<double>': 'doublelist',
+    'IntArrayRef': 'intlist',
+    'Scalar': 'scalar',
+    'ScalarType': 'scalartype',
+    'Dimname': 'dimname',
+    'DimnameList': 'dimnamelist',
+    'TensorList': 'tensorlist',
+    'int64_t': 'toInt64',
+    'bool': 'toBool',
+    'double': 'toDouble',
+    'std::string': 'string',
+    'c10::optional<std::string>': 'stringOptional',
+}
+
+UNPACK_WITH_SIZE_METHODS = {
+    'TensorList': 'tensorlist_n<{}>',
+    'DimnameList': 'dimnamelist',
+    'IntArrayRef': 'intlist',
+    'c10::optional<IntArrayRef>': 'intlistOptional',
+}
+
+UNPACK_WITH_DEFAULT_METHODS = {
+    'const ScalarType &': 'scalartypeWithDefault',
+    'const Device &': 'deviceWithDefault',
+    'c10::optional<Layout>': 'layoutWithDefault',
+}
+
+# Return RHS expression for python argument using PythonArgParser output.
+# e.g. for arg name 'foo', arg type 'bool', arg_index = 2, returns '_r.toBool(2)'
+def arg_parser_output_expr(
+    arg_index: int, a: PythonArgument, la: Optional[DispatchLambdaArgument]
+) -> PythonArgParserOutputExpr:
+    # The same python signature (and python schema string) is usually
+    # associated with two aten C++ functions: the base version and the
+    # out-place variant. Usually the two functions have the same set of
+    # arguments - of course, except for the output arguments. But in some
+    # cases they might have slightly different C++ argument types -
+    # affected by the 'use_c10_dispatcher' state.
+    #
+    # More specially, 'Tensor?' type can be translated into
+    # either 'const c10::optional<Tensor>&' or 'const Tensor &'.
+    # Unfortunately, this difference can affect how we should access arg
+    # parser output. The former expects '_r.optionalTensor(i)' while the
+    # latter expects '_r.tensor(i)'.
+    #
+    # Because of this subtle difference, we cannot solely use the shared
+    # python signature to determine the RHS expr for both C++ variants.
```

Comment:
The difference is determined by whether the op is c10_full or not though, right? And the result of that difference is inherited here via the lambda arg. (And the check of the arg name actually disables the use of the lamdba arg type, the original source of the difference being arcane lifetime reasons IIRC.)

Just checking my understanding - at first I thought this was implying that the type differences were due multiplexing the out-place variants through here, but on second read I think it's simply describing both influences at once.

Reply:
Right, I think it's the combination of (out-place, c10_full):
c10_full alone already makes it hard for us to differentiate _r.tensor(i) v.s. _r.optionalTensor(i) cases solely based upon the python signature (as they both share the same python argument type 'Tensor?').
But we do have 'cpp_type_str' stored inside each python argument (which is copied from cpp signature and respects the c10_full state). So we could use the cpp_type_str to choose between _r.tensor v.s. _r.optionalTensor had it not been for the out-place multiplexing.

Seems in most cases base and out-place have the same c10_full state. There are only a couple cases where base and out-place have different c10_full state, which forced it to use different '_r.tensor(i)' v.s. '_r.optionalTensor(i)' to read the same PythonArgOutput field for base and out-place variant, respectively.

Hmm, shall we create separate PythonSignature instances for base and out-place variants? Still, I feel the real difference is on the lambda arg side instead of the python signature. The lambda arg type is the dominating factor so we should explicitly pass in it.

Reply:
Whew, this is probably the worst part of the old codegen logic. Let me suggest an alternative possibility, which is that you replicate the `use_c10_dispatcher` structure from tools.codegen.api.cpp when constructing PythonArgument (or, perhaps more precisely, when you construct the binding between PythonArgument and CppArgument).

Right now, you make use of the cpp_type_str (which is indirectly produced by argumenttype_type) to reuse the computation of whether or not you are in tensor or optionalTensor. But this reuse is kind of brittle, because the returned C++ type is not structured at all, and so you are having to reconstruct information e.g., via `UNPACK_METHODS ` which adds a bunch of new invariants on how you can refactor `argumenttype_type` in the future; if I change how I spell a C++ type, there's a nonlocal effect that the type checker won't tell me about where I have to go update some map somewhere else.

An alternate way to write this code is to literally take the code at `argumenttype_type`, copy paste it to this file, and then modify it so that you produce the correct unpacking calls. This code would then replace `UNPACK_METHODS`. It would look something like:

```
    if isinstance(t, BaseType):
        if t.name == BaseTy.Tensor:
            return 'tensor'
        ... # need cases for other base types
        else:
            raise AssertionError(f"base type should have been value type {t}")
    elif isinstance(t, OptionalType):
        if str(t.elem) == 'Tensor':
            if mutable:
                return 'tensor'
            else:
                if local.use_c10_dispatcher().dispatcher_uses_new_style():
                    return 'optionalTensor'
                else:
                    return 'tensor'
```

Pros of this method are that you can now completely decouple Python code generation from cpp api. The con is that you have to update this logic when you update cpp api, but this shouldn't be surprising at all: *of course* you have to change the python-to-cpp translation code when you change the cpp code.

WDYT?

Reply:
Added TODO to emit UNPACK_METHODS directly from Type and avoid indirect translation via cpp_type_str.